File: app/src/main/java/eu/kanade/mangafeed/ui/setting/preference/LoginDialogPreference.java
package eu.kanade.mangafeed.ui.setting.preference;

import android.app.AlertDialog;
import android.content.Context;
import android.os.Bundle;
import android.preference.DialogPreference;
import android.text.method.PasswordTransformationMethod;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;

import com.dd.processbutton.iml.ActionProcessButton;

import butterknife.Bind;
import butterknife.ButterKnife;
import eu.kanade.mangafeed.R;
import eu.kanade.mangafeed.data.preference.PreferencesHelper;
import rx.Subscription;

public abstract class LoginDialogPreference extends DialogPreference {

    @Bind(R.id.accounts_login) TextView title;
    @Bind(R.id.username) EditText username;
    @Bind(R.id.password) EditText password;
    @Bind(R.id.show_password) CheckBox showPassword;
    @Bind(R.id.login) ActionProcessButton loginBtn;

    protected PreferencesHelper preferences;
    protected AlertDialog dialog;
    protected Subscription requestSubscription;
    protected Context context;

    public LoginDialogPreference(Context context, PreferencesHelper preferences) {
        super(context, null);
        this.context = context;
        this.preferences = preferences;

        setDialogLayoutResource(R.layout.pref_account_login);
    }

    @Override
    protected void onPrepareDialogBuilder(AlertDialog.Builder builder) {
        // Hide positive button
        builder.setPositiveButton("", this);
    }

    @Override
    protected void onBindDialogView(View view) {
        super.onBindDialogView(view);
        ButterKnife.bind(this, view);

        showPassword.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked)
                password.setTransformationMethod(null);
            else
                password.setTransformationMethod(new PasswordTransformationMethod());
        });

        loginBtn.setMode(ActionProcessButton.Mode.ENDLESS);
        loginBtn.setOnClickListener(click -> checkLogin());
    }

    @Override
    public void showDialog(Bundle state) {
        super.showDialog(state);
        dialog = ((AlertDialog) getDialog());
    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        if (requestSubscription != null)
            requestSubscription.unsubscribe();
    }

    protected abstract void checkLogin();

}

File: app/src/main/java/eu/kanade/tachiyomi/widget/preference/LoginDialogPreference.kt
package eu.kanade.tachiyomi.widget.preference

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import androidx.annotation.StringRes
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.customView
import com.bluelinelabs.conductor.ControllerChangeHandler
import com.bluelinelabs.conductor.ControllerChangeType
import com.dd.processbutton.iml.ActionProcessButton
import eu.kanade.tachiyomi.R
import eu.kanade.tachiyomi.data.preference.PreferencesHelper
import eu.kanade.tachiyomi.databinding.PrefAccountLoginBinding
import eu.kanade.tachiyomi.ui.base.controller.DialogController
import uy.kohesive.injekt.injectLazy

abstract class LoginDialogPreference(
    @StringRes private val usernameLabelRes: Int? = null,
    bundle: Bundle? = null
) : DialogController(bundle) {

    var binding: PrefAccountLoginBinding? = null
        private set

    val preferences: PreferencesHelper by injectLazy()

    override fun onCreateDialog(savedViewState: Bundle?): Dialog {
        binding = PrefAccountLoginBinding.inflate(LayoutInflater.from(activity!!))
        val titleName = activity!!.getString(getTitleName())
        val dialog = MaterialDialog(activity!!)
            .title(text = activity!!.getString(R.string.login_title, titleName))
            .customView(view = binding!!.root)
            .negativeButton(android.R.string.cancel)

        onViewCreated(dialog.view)

        return dialog
    }

    fun onViewCreated(view: View) {
        if (usernameLabelRes != null) {
            binding!!.usernameLabel.hint = view.context.getString(usernameLabelRes)
        }

        binding!!.login.setMode(ActionProcessButton.Mode.ENDLESS)
        binding!!.login.setOnClickListener { checkLogin() }

        setCredentialsOnView(view)
    }

    override fun onChangeStarted(handler: ControllerChangeHandler, type: ControllerChangeType) {
        super.onChangeStarted(handler, type)
        if (!type.isEnter) {
            onDialogClosed()
        }
    }

    open fun onDialogClosed() {
        binding = null
    }

    @StringRes
    protected abstract fun getTitleName(): Int

    protected abstract fun checkLogin()

    protected abstract fun setCredentialsOnView(view: View)
}
